src/hotspot/share/runtime/handshake.cpp

```diff
@@ -565,39 +567,31 @@ HandshakeState::ProcessResult HandshakeState::try_process(HandshakeOperation* ma
 
   Thread* current_thread = Thread::current();
 
-  HandshakeState::ProcessResult pr_ret = HandshakeState::_processed;
-  int executed = 0;
+  HandshakeOperation* op = get_op();
 
-  do {
-    HandshakeOperation* op = pop();
-    if (op != NULL) {
-      assert(SafepointMechanism::local_poll_armed(_handshakee), "Must be");
-      assert(op->_target == NULL || _handshakee == op->_target, "Wrong thread");
-      log_trace(handshake)("Processing handshake " INTPTR_FORMAT " by %s(%s)", p2i(op),
-                           op == match_op ? "handshaker" : "cooperative",
-                           current_thread->is_VM_thread() ? "VM Thread" : "JavaThread");
-
-      if (op == match_op) {
-        pr_ret = HandshakeState::_succeeded;
-      }
+  assert(op != NULL, "Must have an op");
+  assert(SafepointMechanism::local_poll_armed(_handshakee), "Must be");
+  assert(op->_target == NULL || _handshakee == op->_target, "Wrong thread");
 
-      op->prepare(_handshakee, current_thread);
+  log_trace(handshake)("Processing handshake " INTPTR_FORMAT " by %s(%s)", p2i(op),
+                       op == match_op ? "handshaker" : "cooperative",
+                       current_thread->is_VM_thread() ? "VM Thread" : "JavaThread");
 
-      _active_handshaker = current_thread;
-      op->do_handshake(_handshakee);
-      _active_handshaker = NULL;
+  op->prepare(_handshakee, current_thread);
 
-      executed++;
-    }
-  } while (have_non_self_executable_operation());
```

Comment:
Not at all clear how/why this looping can now be removed.

Reply:
It must be removed.
Since we have cocnurrent adds.
Thread P: get OP X
Thread P: do OP X
Thread P: remove OP X
Handshakee: is queue empty? Yes, can continue to e.g. Java
Thread X: add OP Z
Thread P: Is there more OP's to be execute ? Yes => Error, handshakee did not see that OP.

So Thread P needs to verify that the handshakee did see that operation to making sure poll is armed and that the handshakee is in handshake safe state. Which means we need to redo everything, by just returning from this method we do this.